expat/fuzz/xml_lpm_fuzzer.cpp

```diff
@@ -0,0 +1,432 @@
+      encoding = "US-ASCII";
+      break;
+
+    case xml_lpm_fuzzer::Encoding::NONE:
+      encoding = NULL;
+      break;
+
+    default:
+      encoding = "UNKNOWN";
+      break;
+  }
+}
+
+static int allocation_count = 0;
+static std::vector<int> fail_allocations = {};
+
+void* MallocHook(size_t size) {
+  for (auto index : fail_allocations) {
+    if (index == allocation_count) {
+      return NULL;
+    }
+  }
+  allocation_count += 1;
+  return malloc(size);
+}
+
+void* ReallocHook(void* ptr, size_t size) {
+  for (auto index : fail_allocations) {
+    if (index == allocation_count) {
+      return NULL;
+    }
+  }
+  allocation_count += 1;
+  return realloc(ptr, size);
+}
+
+void FreeHook(void* ptr) {
+  free(ptr);
+}
+
+XML_Memory_Handling_Suite memory_handling_suite = {
+  MallocHook, ReallocHook, FreeHook
+};
+
+void InitializeParser(XML_Parser parser);
+
+// We want a parse function that supports resumption, so that we can cover the
+// suspend/resume code.
+enum XML_Status Parse(XML_Parser parser, const XML_Char* input, int input_len,
+                      int is_final) {
+  enum XML_Status status = XML_Parse(parser, input, input_len, is_final);
+  while (status == XML_STATUS_SUSPENDED) {
+    status = XML_ResumeParser(parser);
+  }
+  return status;
+}
+
+// When the fuzzer is compiled with instrumentation such as ASan, then the
+// accesses in TouchString will fault if they access invalid memory (ie. detect
+// either a use-after-free or buffer-overflow). By calling TouchString in each
+// of the callbacks, we can check that the arguments meet the API specifications
+// in terms of length/null-termination. no_optimize is used to ensure that the
+// compiler has to emit actual memory reads, instead of removing them.
+static volatile size_t no_optimize = 0;
+static void TouchString(const XML_Char* ptr, int len=-1) {
+  if (!ptr) {
+    return;
+  }
+
+  if (len == -1) {
+    for (XML_Char value = *ptr++; value; value = *ptr++) {
+      no_optimize += value;
+    }
+  } else {
+    for (int i = 0; i < len; ++i) {
+      no_optimize += ptr[i];
+    }
+  }
+}
+
+static void TouchChildNodes(XML_Content* content, bool topLevel=true) {
+  switch (content->type) {
+    case XML_CTYPE_EMPTY:
+    case XML_CTYPE_ANY:
+      assert(content->quant == XML_CQUANT_NONE);
+      assert(content->name == NULL);
+      assert(content->numchildren == 0);
+      assert(content->children == NULL);
+      break;
+
+    case XML_CTYPE_MIXED:
+      assert(content->quant == XML_CQUANT_NONE
+             || content->quant == XML_CQUANT_REP);
+      assert(content->name == NULL);
+      for (int i = 0; i < content->numchildren; ++i) {
+        assert(content->children[i].type == XML_CTYPE_NAME);
+        assert(content->children[i].numchildren == 0);
+        TouchString(content->children[i].name);
+      }
+      break;
+
+    case XML_CTYPE_NAME:
+      assert(content->numchildren == 0);
+      TouchString(content->name);
```

Comment:
Let's add:
```c
assert(content->children == NULL);
```
On another note, `content->quant` could be any of the four valid `XML_QUANT_*` values here. If it's not too verbose, we could add e.g.:
```c
assert((content->quant == XML_CQUANT_NONE) || (content->quant == XML_CQUANT_OPT) || (content->quant == XML_CQUANT_REP) || (content->quant == XML_CQUANT_PLUS));
```